flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -0,0 +1,417 @@
+					taskName,
+					barrierId,
+					currentCheckpointId);
+
+				// this stops the current alignment
+				releaseBlocksAndResetBarriers();
+
+				// the next checkpoint starts as canceled
+				currentCheckpointId = barrierId;
+				threadSafeUnaligner.setCurrentCheckpointId(currentCheckpointId);
+				notifyAbortOnCancellationBarrier(barrierId);
+			}
+
+			// else: ignore trailing (cancellation) barrier from an earlier checkpoint (obsolete now)
+
+		}
+		else if (barrierId > currentCheckpointId) {
+			// first barrier of a new checkpoint is directly a cancellation
+
+			// by setting the currentCheckpointId to this checkpoint while keeping the numBarriers
+			// at zero means that no checkpoint barrier can start a new alignment
+			currentCheckpointId = barrierId;
+			threadSafeUnaligner.setCurrentCheckpointId(currentCheckpointId);
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("{}: Checkpoint {} canceled, skipping alignment.", taskName, barrierId);
+			}
+
+			notifyAbortOnCancellationBarrier(barrierId);
+		}
+
+		// else: trailing barrier from either
+		//   - a previous (subsumed) checkpoint
+		//   - the current checkpoint if it was already canceled
+		return false;
+	}
+
+	@Override
+	public boolean processEndOfPartition() throws Exception {
+		threadSafeUnaligner.processEndOfPartition();
+
+		if (numBarrierConsumed > 0) {
+			// let the task know we skip a checkpoint
+			notifyAbort(
+				currentCheckpointId,
+				new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED_INPUT_END_OF_STREAM));
+			// no chance to complete this checkpoint
+			releaseBlocksAndResetBarriers();
+			return true;
+		}
+		return false;
+	}
+
+	@Override
+	public long getLatestCheckpointId() {
+		return currentCheckpointId;
+	}
+
+	@Override
+	public long getAlignmentDurationNanos() {
+		return 0;
+	}
+
+	@Override
+	public String toString() {
+		return String.format("%s: last checkpoint: %d", taskName, currentCheckpointId);
+	}
+
+	@Override
+	public void checkpointSizeLimitExceeded(long maxBufferedBytes) {
+	}
+
+	@Override
+	public void close() throws IOException {
+		super.close();
+		threadSafeUnaligner.close();
+	}
+
+	@Override
+	protected boolean isBarrierConsumed(long checkpointId, int channelIndex) {
+		if (checkpointId < currentCheckpointId) {
+			return true;
+		}
+		if (checkpointId > currentCheckpointId) {
+			return false;
+		}
+		return barrierConsumedChannels[channelIndex];
+	}
+
+	public CompletableFuture<Void> getAllBarriersReceivedFuture(long checkpointId) {
+		return threadSafeUnaligner.getAllBarriersReceivedFuture(checkpointId);
+	}
+
+	@Override
+	public Optional<BufferReceivedListener> getBufferReceivedListener() {
+		return Optional.of(threadSafeUnaligner);
+	}
+
+	private int getFlattenedChannelIndex(InputChannelInfo channelInfo) {
+		return gateChannelOffsets[channelInfo.getGateIdx()] + channelInfo.getInputChannelIdx();
+	}
+
+	private static class ThreadSafeUnaligner implements BufferReceivedListener, Closeable {
```

Comment:
Annotated and I renamed all members with the same name to make it more clear. I'd like to avoid moving it out, since it's basically two sides of the same coin for me.

Reply:
It implements an interface that is used by classes outside of `CheckpointBarrierUnaligner`. So to me, it's a separate entity. But it's quite subjective so feel free to ignore it.